yarn-project/noir-protocol-circuits-types/src/scripts/generate_ts_from_abi.ts

```diff
@@ -102,99 +65,217 @@ function getLastComponentOfPath(str: string): string {
+  /** All the types in the ABIs */
+  private allTypes: AbiTypeWithGenerics[] = [];
+  /** The demonomorphized ABIs of the circuits */
+  private demonomorphizedAbis: {
+    circuitName: string;
+    params: { name: string; type: AbiTypeWithGenerics }[];
+    returnType?: AbiTypeWithGenerics;
+  }[] = [];
+  /** Maps struct id to name for structs with the same name and different field sets */
+  private structIdToTsName = new Map<string, string>();
+  /** Collect all the primitives used in the types to add them to the codegen */
+  private primitiveTypesUsed = new Map<string, PrimitiveTypesUsed>();
+
+  constructor(circuits: { abi: NoirFunctionAbi; circuitName: string }[]) {
+    // Map all the types used in the ABIs to the demonomorphized types
+    for (const { abi, circuitName } of circuits) {
+      const params = abi.parameters.map(param => {
+        const type = mapAbiTypeToAbiTypeWithGenerics(param.type);
+        this.allTypes.push(type);
+        return { name: param.name, type };
+      });
+      if (abi.return_type) {
+        const returnType = mapAbiTypeToAbiTypeWithGenerics(abi.return_type.abi_type);
+        this.allTypes.push(returnType);
+        this.demonomorphizedAbis.push({ circuitName, params, returnType });
+      } else {
+        this.demonomorphizedAbis.push({ circuitName, params });
+      }
+    }
+    // Demononmorphize the types
+    Demonomorphizer.demonomorphize(this.allTypes);
   }
 
-  // List of structs encountered while viewing this type that we need to generate
-  // bindings for.
-  const typesEncountered = new Set<AbiType>();
-
-  // Codegen the struct and then its fields, so that the structs fields
-  // are defined before the struct itself.
-  let codeGeneratedStruct = '';
-  let codeGeneratedStructFields = '';
-
-  const structName = getLastComponentOfPath(type.path);
-  if (!output.has(structName)) {
-    codeGeneratedStruct += `export interface ${structName} {\n`;
-    for (const field of type.fields) {
-      codeGeneratedStruct += `  ${field.name}: ${abiTypeToTs(field.type)};\n`;
-      typesEncountered.add(field.type);
+  public codegen(): string {
+    this.primitiveTypesUsed = new Map<string, PrimitiveTypesUsed>();
+    const structsCode = this.codegenAllStructs();
+    const interfacesCode = this.codegenAllInterfaces();
+    const primitivesCode = this.codegenAllPrimitives();
+
+    return `
+    /* Autogenerated file, do not edit! */
+    /* eslint-disable */
+    ${primitivesCode}
+    ${structsCode}
+    ${interfacesCode}`;
+  }
+
+  private codegenAllStructs(): string {
+    const allStructs = this.allTypes.flatMap(findAllStructsInType);
+    // First, deduplicate the structs used
+    const structTypesToExport = new Map<string, StructType>();
+    for (const struct of allStructs) {
+      const id = Demonomorphizer.buildIdForStruct(struct.structType);
+      if (structTypesToExport.has(id)) {
+        continue;
+      }
+      structTypesToExport.set(id, struct.structType);
```

Comment:
Since the types used are passed through the demonomorphizer, it guarantees that it'll only generate one struct type for all struct variants with the same id. In this case, since these two structs have the same name and field names, it'll demonomorphize them into
```rust
struct Foo<A> {
  value: A
}
```
So the generate_ts_from_abi can safely generate one type for each id

Reply:
Nice!